src/interp.rs

```diff
@@ -19,153 +19,483 @@
 
-        Interp {
-            factor,
-            taps,
-            channels,
-            delay,
-            filter: filter
-                .into_iter()
-                .map(|f| Filter {
-                    coeff: f.into_boxed_slice(),
-                })
-                .collect(),
-            z: z.into_boxed_slice(),
-            zi: 0,
+        pub fn process(&mut self, src: &[f32], dst: &mut [f32]) {
+            assert!(src.len().checked_mul(self.factor) == Some(dst.len()));
+            assert!(self.z.len() == self.delay * self.channels as usize);
+            assert!(self.filter.len() == self.factor);
+            assert!(self.zi < self.delay);
+
+            if src.is_empty() {
+                return;
+            }
+
+            let frames = src.len() / self.channels as usize;
+
+            for (src, (dst, z)) in src.chunks_exact(frames).zip(
+                dst.chunks_exact_mut(frames * self.factor)
+                    .zip(self.z.chunks_exact_mut(self.delay)),
+            ) {
+                let mut zi = self.zi;
+
+                for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
+                    // Add sample to delay buffer
+                    //
+                    // TODO Ringbuffer without bounds checks for z/zi
+                    //
+                    // Safety: zi is checked to be between 0 and self.delay
+                    *unsafe { z.get_unchecked_mut(zi) } = *src;
+
+                    // Apply coefficients
+                    for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
+                        let mut acc = 0.0;
+                        for (c, index) in &*filter.coeff {
+                            let mut i = zi as i32 - *index as i32;
+                            if i < 0 {
+                                i += self.delay as i32;
+                            }
+
+                            // Safety: zi is checked to be between 0 and self.delay
+                            acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
+                        }
+
+                        *dst = acc as f32;
+                    }
+
+                    zi += 1;
+                    if zi == self.delay {
+                        zi = 0;
+                    }
+                }
+            }
+
+            self.zi = (self.zi + frames) % self.delay;
         }
     }
+}
 
-    pub fn get_factor(&self) -> usize {
-        self.factor
+/// A trait to be generic over number of channels in a of frame
+trait Frame: Sized + Copy {
+    const CHANNELS: usize;
+
+    fn scale_add(&mut self, other: &Self, coeff: f32);
+    fn from_planar(slice: &[f32], stride: usize) -> Self;
+}
+
+type MonoFrame32 = [f32; 1];
+type StereoFrame32 = [f32; 2];
+type SurroundFrame32 = [f32; 6];
+
+impl Frame for MonoFrame32 {
+    const CHANNELS: usize = 1;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
```

Comment:
Might be worth using [`f32::mul_add()`](https://doc.rust-lang.org/std/primitive.f32.html#method.mul_add) here. I'd assume the compiler to optimize that accordingly anyway but ...

Reply:
Ooh, nice! Did not know it existed :)

Reply:
Ouchi! I tried that. Unfortunately, and weirdly, it seems to negate almost all the performance improvements of this entire PR. I guess somehow it prevents the auto-vectorization, even though I'd assume it would give the optimizer even better information to work with.

I'll test around a little, I suppose this depends on target CPU-features during compilation

Reply:
Alright. Explicitly targeting '+avx2,+fma' during compilation seem to mostly adsorb the performance-hit of using mul_add. I found some explanations that mul_add as implemented in Rust _require_ the higher-precision-guarantees of mul_add. Not all CPU:s support the SIMD-versions of that, which apparently is why mul_add causes a regression. Not sure what to think of it though. Using `mul_add` might yield better precision, at expense of default-build performance. Ideas/opinions?

Reply:
Some further exploration: Despite using mul_add, the absolute diff with libebu is still just under `0.000002`, no significant improvement seen.

Reply:
Let's ignore that then but can you add a comment in the code about this so a future reader doesn't wonder the same and goes on the same journey? :)

Reply:
Or maybe it can be enabled based on `target_feature` if it makes any difference at all

Reply:
I chose to enable based on a non-default package-feature instead, with the recommendation to also enable the target-feature to avoid performance impact. Basing the accuracy on the hardware compiled for feels a bit counter-intuitive?

For reference, the sum(abs(err)) for non-fma is 4 times bigger than with fma, but the max(abs(err)) is roughly the same.

I think we're satisfied here?